src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,414 @@
+    const auto& original_column =
+        original_table.get_chunk(chunk_id).get_column(reference_column.referenced_column_id());
+    PosListContext pos_list_context = {positions_it, original_positions.cend()};
+
+    // If the ReferenceColumn points to a ValueColumn, perform the ValueColumn scan on the original with a PosList to
+    // determine which values are relevant for the scan.
+    const auto value_column = std::dynamic_pointer_cast<ValueColumn<T>>(original_column);
+    if (value_column) {
+      _execute_scan_on_value_column(*value_column, scan_type, chunk_id, search_value, positions, pos_list_context);
+      continue;
+    }
+
+    // If the ReferenceColumn points to a DictionaryColumn, perform the DictionaryColumn scan on the original with a
+    // PosList to determine which values are relevant for the scan.
+    const auto dictionary_column = std::dynamic_pointer_cast<DictionaryColumn<T>>(original_column);
+    if (dictionary_column) {
+      _execute_scan_on_dictionary_column(*dictionary_column, scan_type, chunk_id, search_value, positions,
+                                         pos_list_context);
+      continue;
+    }
+
+    throw std::logic_error("Unknown column type");
+  }
+}
+
+template <typename T>
+void TableScanDispatcher<T>::_execute_scan_on_value_column(const ValueColumn<T>& value_column, const ScanType scan_type,
+                                                           const ChunkID chunk_id, const T search_value,
+                                                           PosList& positions,
+                                                           std::optional<PosListContext> pos_list_context) {
+  TableScanImpl<T> scan;
+  // We set contains_value to true because we have no way of knowing whether it is in or not, so we cannot optimise.
+  scan.perform_scan(scan_type, chunk_id, value_column.values(), search_value, positions, true, pos_list_context);
+}
+
+template <typename T>
+void TableScanDispatcher<T>::_execute_scan_on_dictionary_column(const DictionaryColumn<T>& dictionary_column,
+                                                                const ScanType scan_type, const ChunkID chunk_id,
+                                                                const T search_value, PosList& positions,
+                                                                std::optional<PosListContext> pos_list_context) {
+  const auto lower_bound = dictionary_column.lower_bound(search_value);
+
+  DebugAssert(dictionary_column.dictionary(), "Dictionary column's dictionary is not set");
+  const auto& dictionary_values = *dictionary_column.dictionary();
+
+  const auto contains_value = lower_bound != INVALID_VALUE_ID && dictionary_values[lower_bound] == search_value;
+
+  // Determine which size attribute vector we have and run the scan on the correct type to avoid virtual calls for
+  // each value access to BaseAttributeVector.
```

Comment:
Does this really improve performance enough to justify the code overhead?

Reply:
yes :)